Implement shared PID namespace in the dockershim

verb opened the pull request:
**What this PR does / why we need it**: Defaults the Docker CRI to using a shared PID namespace for pods. Implements proposal in https://github.com/kubernetes/community/pull/207 tracked by #1615.

//cc @dchen1107 @vishh @timstclair 

**Special notes for your reviewer**: none

**Release note**:
```release-note
Some container runtimes share a process (PID) namespace for all containers in a pod. This will become the default for Docker in a future release of Kubernetes. You can preview this functionality if running with the CRI and Docker 1.13.1 by enabling the --experimental-docker-enable-shared-pid kubelet flag.
```

k8s-reviewable commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.kubernetes.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.kubernetes.io/reviews/kubernetes/kubernetes/41583)
<!-- Reviewable:end -->

timstclair commented:
@k8s-bot ok to test

yujuhong commented on cmd/kubelet/app/options/options.go:
```diff
@@ -260,6 +260,7 @@ func (s *KubeletServer) AddFlags(fs *pflag.FlagSet) {
 
 	fs.StringVar(&s.RemoteRuntimeEndpoint, "container-runtime-endpoint", s.RemoteRuntimeEndpoint, "[Experimental] The unix socket endpoint of remote runtime service. The endpoint is used only when CRI integration is enabled (--enable-cri)")
 	fs.StringVar(&s.RemoteImageEndpoint, "image-service-endpoint", s.RemoteImageEndpoint, "[Experimental] The unix socket endpoint of remote image service. If not specified, it will be the same with container-runtime-endpoint by default. The endpoint is used only when CRI integration is enabled (--enable-cri)")
+	fs.BoolVar(&s.DockerDisableSharedPID, "docker-disable-shared-pid", s.DockerDisableSharedPID, "The Container Runtime Interface (CRI) defaults to using a shared PID namespace for pods. Setting this flag reverts the Docker CRI runtime to the old behavior of isolated pod PID namespaces.")
```
To be clear, s/shared PID namespace for pods/shared PID namespaces for containers in a pod
s/isolated pod PID namespaces/isolated PID namespaces for containers.

yujuhong commented on pkg/apis/componentconfig/types.go:
```diff
@@ -485,6 +485,9 @@ type KubeletConfiguration struct {
 	// This can be useful for debugging volume related issues.
 	KeepTerminatedPodVolumes bool
+	// This flag, if set, disables use of a shared PID namespace for pods running in the docker CRI runtime.
+	// TODO: remove once this escape hatch is no longer required.
```
nit: be more clear when this should happen.

yujuhong commented on pkg/apis/componentconfig/v1alpha1/types.go:
```diff
@@ -522,6 +522,8 @@ type KubeletConfiguration struct {
 	// This can be useful for debugging volume related issues.
 	KeepTerminatedPodVolumes bool `json:"keepTerminatedPodVolumes,omitempty"`
+	// This flag, if set, disables use of a shared PID namespace for pods run by the docker CRI runtime.
+	DockerDisableSharedPID bool `json:"dockerDisableSharedPID,omitempty"`
```
Just in case, make this a pointer (`*bool`) so that we can change the default it needed.

yujuhong commented on pkg/kubelet/dockershim/docker_service.go:
```diff
@@ -236,6 +237,11 @@ type dockerService struct {
+	// This option provides an escape hatch to override the new default behavior for Docker under
+	// the CRI to use a shared PID namespace for all pods. It is temporary and will be removed.
+	// See proposals/pod-pid-namespace.md for details.
+	// TODO: Remove once the escape hatch is no longer used (1.7?)
```
File a github issue for removing this knob and include the issue number here.

yujuhong commented on pkg/kubelet/dockershim/security_context.go:
```diff
@@ -164,9 +166,25 @@ func modifyHostNetworkOptionForContainer(hostNetwork bool, sandboxID string, hc
+	if !strings.HasPrefix(string(hc.PidMode), "container:") {
+		return
+	}
+	if disableSharedPID || (version.Major <= 1 && version.Minor <= 23) {
```
Use the existing helper function for comparing the versions: https://github.com/kubernetes/kubernetes/blob/v1.6.0-beta.0/pkg/kubelet/dockershim/helpers.go#L314

yujuhong commented:
@verb I posted a comment on the original proposal about issues with docker 1.12 + shared pid namespace. I think that's a serious and might block this from being enabled this release.

verb commented:
@yujuhong agreed, but I think the safer move is to bump the minimum docker version for shared PID to 1.13. I don't expect the problems to be addressed in 1.12.

Given that we know of no problems in 1.13 I don't think it should block this from being merged. In fact, I think we should include it to get early feedback from anyone testing with the not-qualified-for-1.6 docker 1.13. It would give us a slower rollout.

yujuhong commented:
@verb, we haven't even started qualifying docker 1.13 for kubernetes. If we don't even list docker 1.13 as a compatible version for k8s 1.6, we wouldn't get enough users to try out this feature. That makes this PR less meaningful (and the changelog/release note unclear). I think we can come back to this in the next release, when we have at least a node e2e image testing docker 1.13+. WDYT?

/cc @dchen1107

verb commented:
The release notes is a good point, but I think that could be worked around. It would probably help to list out the upsides and downsides of merging now:

Upsides of merging now:
- Potentially get early feedback on the feature by people who are explicitly coloring outside of the lines and expect instability.
- Slower rollout for change in default behavior
- Is a no-op for earlier docker versions.
- I don't have to carry forward a patch to an in-flux section of the CRI

Downsides of merging now:
- Difficulty in wording a release note

Perhaps I'm missing something, but I don't understand why we wouldn't just merge this. It seems wasted effort to attempt to align it to releases when it's already aligned with docker versions, and there's only potential upside to including it.

Has 1.12 been qualified yet? I was always kind of expecting we'd skip 1.12 and go right to 1.13 anyway.

vishh commented:
We have users who use more recent versions of docker even though upstream recommends an older version. I agree with @verb's argument that having this PR in v1.6 would give us some early feedback. Perhaps we can make this an `experimental` opt-in feature for v1.6?

yujuhong commented:
I wrote a longer draft response last week but I couldn't find it anywhere :-\

Anyway, my main concern is that enabling this by default without some basic level of confidence on docker 1.13 compatibility may just cause too much noise to have any meaningful feedback. Also, this is a user-facing change, so opt-in may be better than opt-out.

I am against turning this on for docker 1.12 (since it doesn't work), but not so strongly against doing it for 1.13. @dchen1107 could you chime in since the code freeze is imminent?

dchen1107 commented:
I am ok to take this PR to have early engagement with the users if we disable this by default no matter which docker version is since
   * We already know there is an issue in docker 1.12.6. 
   * We didn't do any docker 1.13 validation tests. The existing automated validation pipeline is not sufficient. There is no performance, restart, etc. tests. We didn't run CRI tests against docker 1.13.x yet. 
   * We didn't plan to announce docker 1.13 support in Kubernetes 1.6 release

WDYT? cc/ @yujuhong @verb @Random-Liu

dchen1107 commented on pkg/kubelet/dockershim/security_context.go:
```diff
@@ -163,9 +165,25 @@ func modifyHostNetworkOptionForContainer(hostNetwork bool, sandboxID string, hc
+	if !strings.HasPrefix(string(hc.PidMode), "container:") {
+		return
+	}
+	if disableSharedPID || version.LT(semver.Version{Major: 1, Minor: 25}) {
```
Can we change this one to API version 1.26 (docker 1.13.x)?

verb replied:
ftr, API version 1.25 is docker 1.13.0, but yes, I will bump this to 1.26 which corresponds to docker 1.13.1.

ref: https://docs.docker.com/engine/api/v1.26/

verb commented:
I changed this to be opt-in and reversed the logic of the kubelet flag so it would be more straightforward for users. I left the dockershim implementation to be disable-oriented to keep developers focused on the fact that this will be the default in the near future.

I couldn't find any documentation on how to add an experimental feature, so I just named the flag similarly to how --experimental-cri was named.  The minimum version for shared pid in docker is now 1.13.1.

I think this addresses all comments. If I've missed something please point it out.

verb commented:
I rewrote the release note, but I'm not sure if we add notes for experimental features or not.

verb commented:
AWS e2e failed on trying to tear down previous cluster. Will retry later.

yujuhong commented:
LGTM

verb commented:
@k8s-bot kops aws e2e test this

dchen1107 commented:
/lgtm and
/approve

errordeveloper commented:
I don't know why this was assigned to me, I have very little knowledge of
this area.

On Wed, 1 Mar 2017, 23:09 Kubernetes Submit Queue, <notifications@github.com>
wrote:

> [APPROVALNOTIFIER] This PR is *APPROVED*
>
> The following people have approved this PR: *dchen1107, verb*
>
> Needs approval from an approver in each of these OWNERS Files:
>
>    - cmd/kubelet/OWNERS
>    <https://github.com/kubernetes/kubernetes/blob/master/cmd/kubelet/OWNERS>
>    [dchen1107]
>    - hack/verify-flags/OWNERS
>    <https://github.com/kubernetes/kubernetes/blob/master/hack/verify-flags/OWNERS>
>    [dchen1107]
>    - pkg/OWNERS
>    <https://github.com/kubernetes/kubernetes/blob/master/pkg/OWNERS>
>    [dchen1107]
>
> We suggest the following people:
> cc
> You can indicate your approval by writing /approve in a comment
> You can cancel your approval by writing /approve cancel in a comment
>
> —
> You are receiving this because you were assigned.
> Reply to this email directly, view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/41583#issuecomment-283501622>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAPWS-5hrkNss0dGy_AV1FkItpky-gbkks5rhfqjgaJpZM4MDbKW>
> .
>

yujuhong commented:
> I don't know why this was assigned to me, I have very little knowledge of
> this area.

@errordeveloper probably because the PR touched `pkg/apis` of which you are an owner. The bot is not smart enough to tell what changes in a package are significant.

ethernetdan commented:
Holding off until merging until code freeze is lifted - if this can be classified as a bug please feel free to remove.

/cc @kubernetes/kubernetes-release-managers

calebamiles commented:
Moving to the 1.7 milestone @verb.

**cc: @ethernetdan, @kubernetes/kubernetes-release-managers**

verb committed:
Implement shared PID namespace in the dockershim

verb commented:
squashed and rebased

verb commented:
@k8s-bot unit test this

dchen1107 commented:
/lgtm

verb commented:
@ethernetdan @calebamiles @dchen1107 can we remove do-not-merge now?

verb commented:
rebased again. conflict was upstream removal of `generated/openapi/zz_generated.openapi.go`

yujuhong commented:
/lgtm

verb commented:
@k8s-bot gce etcd3 e2e test this

verb commented:
opened #45133 for flake

@k8s-bot bazel test this
